File: client/sse_test.go
package client

import (
	"bufio"
	"io"
	"strings"
	"testing"
)

func scanFrames(t *testing.T, document string) []frame {
	t.Helper()
	var frames []frame
	err := scanSSE(bufio.NewReader(strings.NewReader(document)), func(f frame) bool {
		frames = append(frames, f)
		return true
	})
	if err != io.EOF {
		t.Fatalf("scan error: %v", err)
	}
	return frames
}

func TestScanSSESimpleFrames(t *testing.T) {
	frames := scanFrames(t, "data: one\n\ndata: two\n\n")
	if len(frames) != 2 {
		t.Fatalf("dispatched %d frames, want 2", len(frames))
	}
	for i, want := range []string{"one", "two"} {
		if frames[i].event != "message" || string(frames[i].data) != want {
			t.Fatalf("frame %d = %q event %q, want message %q", i, frames[i].data, frames[i].event, want)
		}
	}
}

func TestScanSSEMultiLineData(t *testing.T) {
	frames := scanFrames(t, "data: first\ndata: second\ndata:\n\n")
	if len(frames) != 1 {
		t.Fatalf("dispatched %d frames, want 1", len(frames))
	}
	if want := "first\nsecond\n"; string(frames[0].data) != want {
		t.Fatalf("data %q, want %q", frames[0].data, want)
	}
}

func TestScanSSENamedEventAndID(t *testing.T) {
	frames := scanFrames(t, "event: oap-overflow\nid: 42\ndata: {}\n\ndata: next\n\n")
	if len(frames) != 2 {
		t.Fatalf("dispatched %d frames, want 2", len(frames))
	}
	first := frames[0]
	if first.event != "oap-overflow" || first.lastID != "42" || !first.hasID {
		t.Fatalf("first frame %+v", first)
	}
	// Event name and id reset between frames; the default name returns.
	second := frames[1]
	if second.event != "message" || second.hasID {
		t.Fatalf("second frame %+v", second)
	}
}

func TestScanSSECommentsAndKeepalives(t *testing.T) {
	frames := scanFrames(t, ": keepalive\n\ndata: one\n: mid-frame comment\ndata: two\n\n")
	if len(frames) != 1 {
		t.Fatalf("dispatched %d frames, want 1", len(frames))
	}
	if want := "one\ntwo"; string(frames[0].data) != want {
		t.Fatalf("data %q, want %q", frames[0].data, want)
	}
}

func TestScanSSELineTerminators(t *testing.T) {
	document := "data: lf\n\rdata: crlf\r\n\rdata: cr\r\rdata: tail\r\r"
	frames := scanFrames(t, document)
	if len(frames) != 4 {
		t.Fatalf("dispatched %d frames, want 4: %+v", len(frames), frames)
	}
	for i, want := range []string{"lf", "crlf", "cr", "tail"} {
		if string(frames[i].data) != want {
			t.Fatalf("frame %d data %q, want %q", i, frames[i].data, want)
		}
	}
}

func TestScanSSELeadingBOM(t *testing.T) {
	frames := scanFrames(t, "\xEF\xBB\xBFdata: one\n\n")
	if len(frames) != 1 || string(frames[0].data) != "one" {
		t.Fatalf("frames %+v", frames)
	}
}

func TestScanSSEFieldRules(t *testing.T) {
	// One optional space after the colon is dropped, further spaces are kept;
	// a field line without a colon names an unknown field, which is ignored
	// alongside retry and anything unrecognized.
	frames := scanFrames(t, "data:  two spaces\ndata:one\nretry: 100\nunknown: x\nnosolondata\n\n")
	if len(frames) != 1 {
		t.Fatalf("dispatched %d frames, want 1", len(frames))
	}
	if want := " two spaces\none"; string(frames[0].data) != want {
		t.Fatalf("data %q, want %q", frames[0].data, want)
	}
}

func TestScanSSEIDWithNULDiscarded(t *testing.T) {
	frames := scanFrames(t, "id: a\x00b\ndata: one\n\nid: 7\ndata: two\n\n")
	if len(frames) != 2 {
		t.Fatalf("dispatched %d frames, want 2", len(frames))
	}
	if frames[0].hasID {
		t.Fatalf("id containing NUL must be discarded: %+v", frames[0])
	}
	if !frames[1].hasID || frames[1].lastID != "7" {
		t.Fatalf("second frame id: %+v", frames[1])
	}
}

func TestScanSSENoDataNoDispatch(t *testing.T) {
	// Blank lines and id-only frames dispatch nothing; data: with an empty
	// value still dispatches an empty payload.
	frames := scanFrames(t, "\n\nid: 1\n\ndata:\n\n")
	if len(frames) != 1 {
		t.Fatalf("dispatched %d frames, want 1", len(frames))
	}
	if len(frames[0].data) != 0 {
		t.Fatalf("data %q, want empty", frames[0].data)
	}
}

func TestScanSSEUnterminatedTailDiscarded(t *testing.T) {
	frames := scanFrames(t, "data: one\n\ndata: two\n")
	if len(frames) != 1 || string(frames[0].data) != "one" {
		t.Fatalf("frames %+v", frames)
	}
}

func TestScanSSEStopsWhenHandlerDeclines(t *testing.T) {
	// The scan stops at the first declined frame and reads no further, so
	// scanning the same reader again resumes with the following frame.
	reader := bufio.NewReader(strings.NewReader("data: one\n\ndata: two\n\ndata: three\n\n"))
	var first frame
	if err := scanSSE(reader, func(f frame) bool {
		first = f
		return false
	}); err != nil {
		t.Fatalf("scan error: %v", err)
	}
	if string(first.data) != "one" {
		t.Fatalf("first frame %q", first.data)
	}
	var second frame
	if err := scanSSE(reader, func(f frame) bool {
		second = f
		return false
	}); err != nil {
		t.Fatalf("scan error: %v", err)
	}
	if string(second.data) != "two" {
		t.Fatalf("second frame %q", second.data)
	}
}

File: clients/ts/test/sse.test.ts
/**
 * Parser unit tests, one per WHATWG rule that matters on this wire — the
 * TypeScript port of the Go client's sse_test.go scenarios.
 */

import assert from 'node:assert/strict';
import { test } from 'node:test';
import { SSEParser, type SSEFrame } from '../src/sse.js';

const encoder = new TextEncoder();

/** Parses a whole document in one chunk the way the daemon writes them. */
function scanFrames(document: string): SSEFrame[] {
  const parser = new SSEParser();
  const frames = parser.push(encoder.encode(document));
  frames.push(...parser.finish());
  return frames;
}

/** Parses a document one byte at a time: every rule must survive chunk splits. */
function scanFramesByteWise(document: string): SSEFrame[] {
  const parser = new SSEParser();
  const bytes = encoder.encode(document);
  const frames: SSEFrame[] = [];
  for (const byte of bytes) frames.push(...parser.push(new Uint8Array([byte])));
  frames.push(...parser.finish());
  return frames;
}

function data(frame: SSEFrame): string {
  return frame.data;
}

test('simple frames dispatch at their blank lines', () => {
  const frames = scanFrames('data: one\n\ndata: two\n\n');
  assert.equal(frames.length, 2);
  for (const [index, want] of ['one', 'two'].entries()) {
    assert.equal(frames[index].event, 'message');
    assert.equal(data(frames[index]), want);
  }
});

test('multiple data lines join with newlines', () => {
  const frames = scanFrames('data: first\ndata: second\ndata:\n\n');
  assert.equal(frames.length, 1);
  assert.equal(data(frames[0]), 'first\nsecond\n');
});

test('named events and ids set, then reset between frames', () => {
  const frames = scanFrames('event: oap-overflow\nid: 42\ndata: {}\n\ndata: next\n\n');
  assert.equal(frames.length, 2);
  const first = frames[0];
  assert.equal(first.event, 'oap-overflow');
  assert.equal(first.lastId, '42');
  assert.ok(first.hasId);
  // Event name and id reset between frames; the default name returns.
  const second = frames[1];
  assert.equal(second.event, 'message');
  assert.ok(!second.hasId);
});

test('comments and keepalives are ignored', () => {
  const frames = scanFrames(': keepalive\n\ndata: one\n: mid-frame comment\ndata: two\n\n');
  assert.equal(frames.length, 1);
  assert.equal(data(frames[0]), 'one\ntwo');
});

test('CR, LF, and CRLF all terminate lines', () => {
  const document = 'data: lf\n\rdata: crlf\r\n\rdata: cr\r\rdata: tail\r\r';
  const frames = scanFrames(document);
  assert.equal(frames.length, 4);
  for (const [index, want] of ['lf', 'crlf', 'cr', 'tail'].entries()) {
    assert.equal(data(frames[index]), want);
  }
});

test('a leading UTF-8 BOM is stripped', () => {
  const frames = scanFrames('﻿data: one\n\n');
  assert.equal(frames.length, 1);
  assert.equal(data(frames[0]), 'one');
});

test('field rules: one optional space dropped, further spaces kept; unknown fields ignored', () => {
  const frames = scanFrames('data:  two spaces\ndata:one\nretry: 100\nunknown: x\nnosolondata\n\n');
  assert.equal(frames.length, 1);
  assert.equal(data(frames[0]), ' two spaces\none');
});

test('an id containing NUL is discarded', () => {
  const frames = scanFrames('id: a\0b\ndata: one\n\nid: 7\ndata: two\n\n');
  assert.equal(frames.length, 2);
  assert.ok(!frames[0].hasId, 'id containing NUL must be discarded');
  assert.ok(frames[1].hasId);
  assert.equal(frames[1].lastId, '7');
});

test('no data means no dispatch, but empty data still dispatches', () => {
  const frames = scanFrames('\n\nid: 1\n\ndata:\n\n');
  assert.equal(frames.length, 1);
  assert.equal(data(frames[0]), '');
});

test('an unterminated trailing frame is discarded', () => {
  const frames = scanFrames('data: one\n\ndata: two\n');
  assert.equal(frames.length, 1);
  assert.equal(data(frames[0]), 'one');
});

test('every rule survives byte-at-a-time chunking', () => {
  const document =
    '﻿event: oap-overflow\nid: 42\rdata: multi\r\ndata: line\ndata:\n: comment\nretry: 9\n\nid: a\0b\ndata: x\n\ndata: tail\n';
  const frames = scanFramesByteWise(document);
  assert.equal(frames.length, 2);
  assert.equal(frames[0].event, 'oap-overflow');
  assert.equal(frames[0].lastId, '42');
  assert.equal(data(frames[0]), 'multi\nline\n');
  assert.ok(!frames[1].hasId);
  assert.equal(data(frames[1]), 'x');
});

test('multi-byte UTF-8 split across chunks decodes correctly', () => {
  const parser = new SSEParser();
  const bytes = encoder.encode('data: héllo ✓\n\n');
  const split = 8; // inside the é sequence
  const first = parser.push(bytes.slice(0, split));
  const second = parser.push(bytes.slice(split));
  parser.finish();
  assert.equal(first.length, 0);
  assert.equal(second.length, 1);
  assert.equal(second[0].data, 'héllo ✓');
});
